objects/src/transaction/proven_tx.rs

```diff
@@ -159,6 +125,202 @@ impl ProvenTransaction {
+    initial_account_hash: Digest,
+
+    /// The hash of the account after the transaction was executed.
+    final_account_hash: Digest,
+
+    /// State changes to the account due to the transaction.
+    account_details: Option<AccountDetails>,
+
+    /// List of [Nullifier]s of all consumed notes by the transaction.
+    input_notes: Vec<Nullifier>,
+
+    /// List of [NoteEnvelope]s of all notes created by the transaction.
+    output_notes: Vec<NoteEnvelope>,
+
+    /// State of the output notes.
+    output_note_details: BTreeMap<NoteId, Note>,
+
+    /// The script root of the transaction, if one was used.
+    tx_script_root: Option<Digest>,
+
+    /// Block [Digest] of the transaction's reference block.
+    block_ref: Digest,
+
+    /// A STARK proof that attests to the correct execution of the transaction.
+    proof: ExecutionProof,
+}
+
+impl ProvenTransactionBuilder {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a [ProvenTransactionBuilder] used to build a [ProvenTransaction].
+    pub fn new(
+        account_id: AccountId,
+        initial_account_hash: Digest,
+        final_account_hash: Digest,
+        block_ref: Digest,
+        proof: ExecutionProof,
+    ) -> Self {
+        Self {
+            account_id,
+            initial_account_hash,
+            final_account_hash,
+            account_details: None,
+            input_notes: Vec::new(),
+            output_notes: Vec::new(),
+            output_note_details: BTreeMap::new(),
+            tx_script_root: None,
+            block_ref,
+            proof,
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Sets the account's details.
+    pub fn account_details(mut self, account_details: AccountDetails) -> Self {
+        self.account_details = Some(account_details);
+        self
+    }
+
+    /// Add notes consumed by the transaction.
+    pub fn add_input_notes<T>(mut self, notes: T) -> Self
+    where
+        T: IntoIterator<Item = Nullifier>,
+    {
+        self.input_notes.extend(notes);
+        self
+    }
+
+    /// Add notes produced by the transaction.
+    pub fn add_output_notes<T>(mut self, notes: T) -> Self
+    where
+        T: IntoIterator<Item = NoteEnvelope>,
+    {
+        self.output_notes.extend(notes);
+        self
+    }
+
+    /// Add produce notes details.
+    pub fn add_output_note_details<T>(mut self, notes: T) -> Self
+    where
+        T: IntoIterator<Item = (NoteId, Note)>,
+    {
+        self.output_note_details.extend(notes);
+        self
+    }
+
+    /// Set transaction's script root.
+    pub fn tx_script_root(mut self, tx_script_root: Digest) -> Self {
+        self.tx_script_root = Some(tx_script_root);
+        self
+    }
+
+    pub fn build(mut self) -> Result<ProvenTransaction, ProvenTransactionError> {
```

Comment:
I think we should preform the same validation checks as we currently do in the `ProvenTransactionBuilder::build()` method on deserialization as well. The main reason for this is to avoid code duplication (i.e., otherwise, we'd end up with basically the same validation code in multiple places). There are, of course, additional validation checks that we'd need to perform once we have access to additional info - but I think it is OK to perform more "stateless" checks here (e.g., we still want to validate that felts are valid, no extraneous info is attached to the transaction etc.).

One tricky one is validation of the attached proof. Technically, we have all the details here to check if the proof is valid. But it does fill odd to validate the proof on deserialization (and we don't even do it in the `ProvenTransactionBuilder::build()` method).

Let's create an issue to address all of the above in a different PR.

Reply:
@bobbinth I concluded the same when I partly implemented the same validation in block-producer. And when I was thinking about the same validation in the store, I've changed my mind, I think, we should implement this validation in miden-base only once, and we can provide all required information like initial account hashes and proofs as arguments for validation method. And, as @hackaugusto  proposed, we can introduce something like `VerifiedProvenTransaction` which can be constructed only during validation process.

Reply:
Another approach with storage could be providing an interface for needed storage operation, but this would require to make validation in async context, which I would probably try to avoid.

Reply:
> @bobbinth I concluded the same when I partly implemented the same validation in block-producer. And when I was thinking about the same validation in the store, I've changed my mind, I think, we should implement this validation in miden-base only once, and we can provide all required information like initial account hashes and proofs as arguments for validation method. And, as @hackaugusto proposed, we can introduce something like `VerifiedProvenTransaction` which can be constructed only during validation process.

Let's create an issue for this.